test/Runtime/Inputs/custom_rr_abi_utilities.h

```diff
@@ -0,0 +1,111 @@
+  macro( 6, __VA_ARGS__)                                                       \
+  macro( 7, __VA_ARGS__)                                                       \
+  macro( 8, __VA_ARGS__)                                                       \
+  macro( 9, __VA_ARGS__)                                                       \
+  macro(10, __VA_ARGS__)                                                       \
+  macro(11, __VA_ARGS__)                                                       \
+  macro(12, __VA_ARGS__)                                                       \
+  macro(13, __VA_ARGS__)                                                       \
+  macro(14, __VA_ARGS__)                                                       \
+  macro(15, __VA_ARGS__)                                                       \
+  macro(19, __VA_ARGS__)                                                       \
+  macro(20, __VA_ARGS__)                                                       \
+  macro(21, __VA_ARGS__)                                                       \
+  macro(22, __VA_ARGS__)                                                       \
+  macro(23, __VA_ARGS__)                                                       \
+  macro(24, __VA_ARGS__)                                                       \
+  macro(25, __VA_ARGS__)                                                       \
+  macro(26, __VA_ARGS__)                                                       \
+  macro(27, __VA_ARGS__)                                                       \
+  macro(28, __VA_ARGS__)
+
+// Apply `macro` to each function that gets specialized entrypoints. Also pass
+// 1 if the function is a retain variant, and 0 if it's a release variant.
+#define ALL_FUNCTIONS(macro)                                                   \
+  macro(swift_retain, 1)                                                       \
+  macro(swift_release, 0)                                                      \
+  macro(swift_bridgeObjectRetain, 1)                                           \
+  macro(swift_bridgeObjectRelease, 0)
+
+// Emit declarations for variables called xN stored in xN, initialized with
+// regs[N].
+#define PASS_REGS_HELPER(num) \
+  register void *x ## num asm ("x" #num) = regs[num];
+#define PASS_REGS ALL_REGS(PASS_REGS_HELPER)
+
+// Emit an entry in an asm inputs list containing "r" (xN).
+#define REG_INPUTS_HELPER(num) \
+  "r" (x ## num),
+#define REG_INPUTS ALL_REGS(REG_INPUTS_HELPER)
+
+// Make a function called call_function_xN that calls function_xN with registers
+// set to the contents of the given registers array.
+#define MAKE_CALL_FUNC(reg, func)                                              \
+  static inline void call_##func##_x##reg(void **regs) {                       \
+    PASS_REGS                                                                  \
+    asm("bl _" #func "_x" #reg : : REG_INPUTS "i"(0));                         \
+  }
```

Comment:
I wonder if we can ensure that this doesn't setup a frame?  This is also a guaranteed tail call right?  Can we attribute that somehow?

Reply:
I'm not sure it's important, given that this is just for a test. I suppose if you wanted to avoid using the `register ... asm("<reg>")` annotation, you could write this bit entirely in assembly language, in which case it could do a tail call. It couldn't be `inline` then though.

Reply:
We don't actually care about whether this sets up a frame or does a tail call. We will overwrite the frame pointer, but that's OK. Otherwise, we just need it to call the requested runtime function with the registers set to the appropriate values. Whether it does that as a tail or standard call, and with or without a frame, is fine.